protocol/serialization-types-macros/src/auto_from.rs

```diff
@@ -0,0 +1,368 @@
+
+    for target_type in target_types {
+        let ts: TokenStream = quote! {
+            impl ::std::convert::From<#ident> for #target_type {
+                fn from(item: #ident) -> Self {
+                    Self {
+                        #(#field_idents: item.#field_idents.into(),) *
+                        #(#vec_field_idents: item.#vec_field_idents.into_iter().map(::std::convert::Into::into).collect(),) *
+                        #(#option_field_idents: item.#option_field_idents.map(::std::convert::Into::into),) *
+                        #(#boxed_field_idents: ::std::boxed::Box::new((*item.#boxed_field_idents).into()),) *
+                    }
+                }
+            }
+
+            impl ::std::convert::From<#target_type> for #ident {
+                fn from(item: #target_type) -> Self {
+                    Self {
+                        #(#field_idents: item.#field_idents.into(),) *
+                        #(#vec_field_idents: item.#vec_field_idents.into_iter().map(::std::convert::Into::into).collect(),) *
+                        #(#option_field_idents: item.#option_field_idents.map(::std::convert::Into::into),) *
+                        #(#boxed_field_idents: ::std::boxed::Box::new((*item.#boxed_field_idents).into()),) *
+                    }
+                }
+            }
+        }
+        .into();
+        output.extend(ts);
+        output.extend(impl_from_for_versioned(ident, target_type));
+    }
+
+    Some(output)
+}
+
+/// Codegen for structs with unnamed fields
+pub fn auto_from_for_struct_with_unnamed_fields(
+    ident: &proc_macro2::Ident,
+    target_types: &[proc_macro2::TokenStream],
+    named: Punctuated<Field, Token![,]>,
+) -> Option<TokenStream> {
+    let mut pos_token_stream: Vec<proc_macro2::TokenStream> = Vec::new();
+    'outer: for f in named {
+        if f.ident.is_none() {
+            let pos = proc_macro2::Literal::usize_unsuffixed(pos_token_stream.len());
+
+            if let syn::Type::Path(type_path) = f.ty {
+                for seg in type_path.path.segments {
+                    match seg.ident.to_string().as_str() {
+                        "Vec" => {
+                            pos_token_stream.push(
+                                quote! {item.#pos.into_iter().map(::std::convert::Into::into).collect()}
+                            );
+                            continue 'outer;
+                        }
+                        "Option" => {
+                            pos_token_stream
+                                .push(quote! {item.#pos.map(::std::convert::Into::into)});
+                            continue 'outer;
+                        }
+                        "Box" => {
+                            pos_token_stream
+                                .push(quote! {::std::boxed::Box::new((*item.#pos).into())});
+                            continue 'outer;
+                        }
+                        _ => {}
```

Comment:
Would be good to log the ident here, for debugging purposes

Reply:
ditto